lib/charms/sdcore_nms_k8s/v0/fiveg_core_gnb.py

```diff
@@ -0,0 +1,348 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library for the `fiveg_core_gnb` relation.
+
+This library contains the Requires and Provides classes for handling the `fiveg_core_gnb`
+interface.
+
+The purpose of this library is to provide a way for a 5G Core to provide network information and
+configuration to CUs/gNodeBs.
+
+To get started using the library, you need to fetch the library using `charmcraft`.
+
+```shell
+cd some-charm
+charmcraft fetch-lib charms.sdcore_nms_k8s.v0.fiveg_core_gnb
+```
+
+Add the following libraries to the charm's `requirements.txt` file:
+- pydantic
+- pytest-interface-tester
+
+Charms providing the `fiveg_core_gnb` relation should use `FivegCoreGnbProvides`.
+The class `PLMNConfig` represents the configuration of a PLMN for the CU/gNodeB. It is composed by
+the Mobile Country Code (MCC), the Mobile Network Code (MNC), the Slice Service Type (SST) and the
+Slice Differentiator (SD). Each CU can be configured with a single Tracking Area Code (TAC) and
+multiple PLMNS.
+
+Typical usage of this class would look something like:
+
+    ```python
+    ...
+    from charms.sdcore_gnbsim_k8s.v0.fiveg_core_gnb import FivegCoreGnbProvides, PLMNConfig
+    ...
+
+    class SomeProviderCharm(CharmBase):
+
+        def __init__(self, *args):
+            ...
+            self.fiveg_core_gnb_provider = FivegCoreGnbProvides(
+                charm=self,
+                relation_name="fiveg_core_gnb"
+                )
+            ...
+            self.framework.observe(
+                self.on.fiveg_core_gnb_provider_relation_changed,
+                self._on_relation_changed
+            )
+
+        def _on_relation_changed(self, event):
+            ...
+            # implement the logic to populate the list of PLMNs.
+            plmns = [PLMNConfig(mcc=..., mnc=..., sst=..., sd=...)
+            self.fiveg_core_gnb_provider.publish_fiveg_core_gnb_information(
```

Comment:
`on relation changed` we get the information from the remote databag and `on relation joined` we publish the information. 
Even though it's not that important because in our charms we will use the centralized configuration method.

Reply:
I guess the logic itself is fine here. The only problem is when to execute this. My comment above covers that part.

The high level flow is:
1. gNB pushes its name to the data bag
2. NMS charm gets the name and adds the gNB to inventory
3. Operator (user) adds gNB to a network slice
4. `webconsole` fires Pebble notification on the Network Slice change
5. NMS gets the Pebble notification and executes this logic to push the gNB config to the data bag.

Reply:
ohh true, I forgot we'll have the pebble notices so it changes everything.
We need to be extra careful to avoid deadlock here :O
Anyways, I don't think we need to include all this complicated logic to the documentation part. Just an example of how to set and get the information should be enough. WDYT?

Reply:
Agreed. I just put it here as a context/explanation.